### ISSUE TITLE
Question about "attribute matches an attribute match list"

### ISSUE THREAD

**Issue by evilpie** [WICG/sanitizer-api#172]
I am trying to understand mostly step 1 of https://wicg.github.io/sanitizer-api/#attribute-matches-an-attribute-match-list.

>  If attribute’s [local name](https://dom.spec.whatwg.org/#concept-attribute-local-name) does not match the [attribute match list](https://wicg.github.io/sanitizer-api/#attribute-match-list) list’s [key](https://webidl.spec.whatwg.org/#idl-record) and if the key is not "*": Return false.

I am not sure if the current definition really makes sense. The "attributes match list" is not really a list, but more of a map. Shouldn't it be something like like:
> Lookup the attribute's local name in attribute match list,  if such a key doesn't exist lookup "*", ...

(There is probably some other spec we could borrow the appropriate terminology from.)

I am also wondering what happens if we have both an explicit `"*"` star element and something else like `"id"`. Should the `"id"` be preferred to `"*"`, when trying to sanitize an attribute with the local name _id_? (It's also possible that we only allow a single "*" and nothing else in e.g. `allowAttributes`?)

**Comment by otherdaniel** [comment]
On definitions/terms: Yeah, that could be better written. [WhatWG infra](https://infra.spec.whatwg.org/#data-structures) spec usually has the right definitions, as I learned in the mean tim.

Explicit "*" + an actual element: The idea was that if one entry matches, then the list matches. So [ "*", "id" ] matches "id" and anything else, just like ["*"] does. Those two are equivalent.

Disallowing "*" combined with other identifiers (and thus throwing during construction, I guess?) would work, but is arguably less developer-friendly, and maybe an issue with forward compatibility. I don't feel particularly strongly about this aspect, though.

**Comment by evilpie** [comment]
Sorry. I didn't actually understand the second part of your answer so let me give an example.

```javascript
var sanitizer = new Sanitizer({allowAttributes: {"*" : [], "id": ["div"]})
x.setHTML("<div id=abc></div>");
```
In this case if we only lookup `"*"` we don't allow any attribute on any element, but if we instead lookup `"id"` first instead we allow that specific attribute for divs.

**Comment by otherdaniel** [comment]
Ah, I see. The intent is:
- "id" (on "div" element) matches "*", but "div" doesn't match [] => `"*":[]` does not match
- "id" (on "div" element) matches "id", and "div" matches ["div"] => `"id": ["div"]` matches
- since one entry matches the attribute is allowed.

I'll second my previous comment, "that could probably be better written". I still think the description is mostly right, but arguably exceedingly tricky:
- "*sanitize action for an attribute*" step 4 says the attribute passes if *any* list matches
- So, "*attribute matches an attribute match list*" is basically run for every entry in allowAttributes.
- For the first one, `"*": []`, step one passes (because the key is "*"), but step 5 returns false.
- For the second, `"id": ["div"]`, step one passes (because the attribute name matches), step 5, too, because the list contains element name, and step 6 returns true.
- With that, one of the lists matches, and hence *any* list matches is true.

Since, in a reasonable implementation, one probably wouldn't want to keep iterating over the list, one would do two lookups instead: one for the attribute name, and one for "*".

**Comment by evilpie** [comment]
This will be resolved by the spec rewrite in #208.
